File: algorithmique/project-euler/057-squareRootConvergents.cpp
#include <stdio.h>
#include <cmath>
#include <algorithm>
#include <string>
#define MAX 1000*1000*1000

using namespace std;
using ll = unsigned long long int;

struct Fraction {
    ll numer, denom;
    Fraction(ll a, ll b) {
        numer = a;
        denom = b;
    }
    void simplify() {
        ll pgcd = __gcd(numer, denom);
        numer /= pgcd;
        denom /= pgcd;
    }

    void invert() {
        swap(numer, denom);
    }

    bool moreDigitsThanDenom () {
        return (to_string(numer).size() > to_string(denom).size());
    }

    Fraction operator + (const Fraction & other) {
        Fraction res = Fraction(numer*other.denom+denom*other.numer, denom*other.denom);
        return res;
    }
};

int main() {
    int nbExpansions = 1;
    int r = 0;
    Fraction frac = Fraction(3, 2);
    while (nbExpansions <= 1000) {
        frac = frac + Fraction(1, 1);
        frac.invert();
        frac = frac + Fraction(1, 1);
        frac.simplify();
        printf("%llu %llu\n", frac.numer, frac.denom);
        if (frac.moreDigitsThanDenom()) {
            r++;
        }
        while (frac.denom > MAX) {
            frac.denom /= 10;
            frac.numer /= 10;
        }
        nbExpansions++;
    }
    printf("%d\n", r);
}

File: algorithmique/project-euler/057-squareRootConvergents.py
import sys
sys.setrecursionlimit(10000)

def gcd(a, b) :
    return a if b==0 else gcd(b, a%b)

class Fraction :

    def __init__(self, numer, denom) :
        self.numer = numer
        self.denom = denom 

    def invert(self) :
        self.numer, self.denom = self.denom, self.numer

    def simplify(self) :
        pgcd = gcd(self.numer, self.denom)
        self.numer //= pgcd
        self.denom //= pgcd

    def numerMoreDigits (self) :
        return len(str(self.numer)) > len(str(self.denom))
        

    def __add__ (self, other) :
        return Fraction (self.numer*other.denom + self.denom*other.numer, self.denom*other.denom)

def main() :
    f = Fraction(3, 2)
    i = 1
    result = 0
    while i <= 1000 :
        f += Fraction(1, 1)
        f.invert()
        f += Fraction(1, 1)
        f.simplify()
        # print(f.numer, f.denom)
        if f.numerMoreDigits() :
            result += 1
        i += 1
    print(result)

main()